### PR TITLE
refactor: Tokenserver: Add mature MySQL adapter

### LINKED ISSUE #1054: Tokenserver: Build more mature database adapter
We'll need to add support for a second, dedicated MySQL database instance for Tokenserver. The necessary infrastructure will likely live in`src/tokenserver/db/mysql`, but more investigation may be needed.

### PR DESCRIPTION
**Pull request by ethowitz**
## Description

Adds mature infrastructure for a Tokenserver-specific MySQL database.

## How should reviewers test?
* Create an account on FxA stage: https://accounts.stage.mozaws.net/signup
* Set `identity.fxaccounts.remote.oauth.uri` to https://oauth.stage.mozaws.net
* Set `identity.fxaccounts.remote.root` to https://accounts.stage.mozaws.net
* Set `identity.fxaccounts.auth.uri` to https://api-accounts.stage.mozaws.net/v1
* Set `identity.fxaccounts.remote.profile.uri` to https://profile.stage.mozaws.net/v1
* Set `identity.sync.tokenserver.uri` to http://localhost:8000/1.0/sync/1.5
* Create a Tokenserver database:
```sql
CREATE DATABASE tokenserver_rs;
```
* Add a service and a node to your database:
```sql
INSERT INTO services (service) VALUES("sync-1.5");
INSERT INTO nodes (service, node, available, capacity) VALUES([service ID from previous step], "http://localhost:8000/", 100, 100);
```
* Set the following values in local.toml:
```toml
enable_tokenserver = true
tokenserver.database_url = "mysql://sample_user:sample_password@localhost/tokenserver_rs"
tokenserver.fxa_oauth_server_url = "https://oauth.stage.mozaws.net"
```
* Add `println!("{}", user_email)` to L47 in `src/tokenserver/handlers.rs`
* Start syncstorage via `make run`
* Restart Firefox, log in to sync, and attempt a sync
* Using the email address from the server logs, add a user to the database:
```sql
INSERT INTO users (service, email, generation, client_state, created_at, nodeid) VALUES([service ID from prior step], "[email address from logs]", 0, "yes", 1620924095, [node ID from prior step]);
```
* Remove the `println!` and recompile/restart the server
* Attempt a sync
* Ensure that the sync was successful via about:sync-log

## Issue(s)

Closes #1054

### CODE REVIEW

**Review by jrconlin — APPROVED** [review]
Nice work on the MockDB stuff.

**Comment by ethowitz** [comment]
Thank you! I'm gonna hold off on merging this for now: there's an issue with how we're parsing Tokenserver settings in `local.toml` that I want to resolve before it lands on master

**Comment by ethowitz** [comment]
@jrconlin I added a default for `tokenserver.database_url` so developers/SRE won't need to configure any Tokenserver settings to develop on/deploy syncstorage-rs (you only need to set Tokenserver settings if you intend to use Tokenserver). I also disabled the Tokenserver route just to be safe.

File: src/tokenserver/migrations/2021-07-16-001122_init/up.sql
Comment (pjenvey) [inline src/tokenserver/migrations/2021-07-16-001122_init/up.sql:1]: I'll note that the services table may be able to be killed entirely.

Tokenserver has solely been used for one service, syncstorage. Either there's always been one entry in it with a pattern of "1.5" or possibly 2 entries (an additional one for pattern "1.1"). cc @rfk if he recalls this

It may make sense to keep it around for now to simplify things, particularly for the eventual migration, but FYI.

File: src/tokenserver/db/models.rs
Comment (pjenvey) [inline src/tokenserver/db/models.rs:80]: We probably want to copy the python version's SQL over directly which will push this sorting into the db layer

File: src/tokenserver/db/models.rs
Comment (pjenvey) [inline src/tokenserver/db/models.rs:66]: FYI: We've tried to standardize the SQL syntax in the rest of the syncstorage impl towards https://www.sqlstyle.guide/ but I won't hold you to it.

File: src/tokenserver/migrations/2021-07-16-001122_init/up.sql
Comment (rfk) [reply src/tokenserver/migrations/2021-07-16-001122_init/up.sql:1]: > Either there's always been one entry in it with a pattern of "1.5" or possibly 2 entries (an additional one for pattern "1.1")

I believe there has only ever been a single entry in production use in this table; references to `1.1` are hangovers from testing and development.

Comment (rfk) [reply src/tokenserver/migrations/2021-07-16-001122_init/up.sql:1]: Just to be very clear: +100 to simplifying the `services` table out of existence entirely

Comment (ethowitz) [reply src/tokenserver/migrations/2021-07-16-001122_init/up.sql:1]: Cool -- I'll add a ticket to remove the `services` table as part of a separate ticket 👍🏻

File: src/tokenserver/db/models.rs
Comment (ethowitz) [reply src/tokenserver/db/models.rs:80]: I think the legacy Tokenserver actually sorts the rows in memory to avoid making the DB indexes too big: https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/assignment/sqlnode/sql.py#L295

Do you think we should change our approach here?

File: src/tokenserver/db/models.rs
Comment (ethowitz) [reply src/tokenserver/db/models.rs:66]: Good to know! I'll update the syntax I added in this PR.

File: src/tokenserver/db/models.rs
Comment (rfk) [reply src/tokenserver/db/models.rs:81]: Two high-level comments on the legacy tokenserver's approach:

* The cost/benefit tradeoff there is entirely theoretical, I don't think we ever measured the potential bloat of indexes or established that it was too expensive.
* The alleged benefit depends on python's `sort` implementation having almost-linear performance on almost-sorted lists.

FWIW I recommend optimizing for simplicity of implementation, whatever that means in the new system.

Comment (ethowitz) [reply src/tokenserver/db/models.rs:81]: I'm inclined to say we should just do the sorting in the code. It seems likely that, in the vast majority of cases, a user will only have a handful of old client states, which is what causes "duplicate" user records to be created. Keeping it in Rust seems to me to be less error-prone than writing it in SQL (though I don't think it matters too much). @pjenvey do you have a strong opinion either way?

Comment (pjenvey) [reply src/tokenserver/db/models.rs:81]: My fault, I was looking solely at the `ORDER BY` clause of the python's get user records call and did not see that it also sorts in code (probably making the former pointless). I'm ok w/ sorting in memory.

**Review by pjenvey — DISMISSED** [review]

**Comment by ethowitz** [comment]
Had to rebase

**Review by jrconlin — APPROVED** [review]

**Pull request merged**
